File: redfox-app/src/main/kotlin/io/github/malczuuu/redfox/app/core/ProjectService.kt
package io.github.malczuuu.redfox.app.core

import io.github.malczuuu.redfox.app.common.Identity
import io.github.malczuuu.redfox.app.common.PageResult
import io.github.malczuuu.redfox.app.domain.ProjectEntity
import io.github.malczuuu.redfox.app.domain.ProjectRepository
import io.github.malczuuu.redfox.app.domain.ThingRepository
import java.time.Clock
import java.time.Instant
import java.util.UUID
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ProjectService(
    private val projectRepository: ProjectRepository,
    private val thingRepository: ThingRepository,
    private val clock: Clock,
) {

  fun getProjects(pageable: Pageable): PageResult<ProjectDto> {
    val projects = projectRepository.findAll(pageable)
    return PageResult(
        content = projects.content.map { it.toDto() },
        page = projects.number,
        size = projects.size,
        totalElements = projects.totalElements,
    )
  }

  fun getProject(id: UUID): ProjectDto {
    val project = projectRepository.findOne(id).orElseThrow { ProjectNotFoundException() }
    return project.toDto()
  }

  @Transactional
  fun createProject(request: CreateProjectDto): Identity {
    if (projectRepository.existsByCode(request.code)) {
      throw ProjectAlreadyExistsException()
    }
    var project =
        ProjectEntity(code = request.code, name = request.name, description = request.description)
    project = projectRepository.save(project)
    return Identity(project.id.toString())
  }

  @Transactional
  fun updateProject(id: UUID, request: UpdateProjectDto) {
    val project = projectRepository.lockOne(id).orElseThrow { ProjectNotFoundException() }
    project.name = request.name
    project.description = request.description
    projectRepository.save(project)
  }

  @Transactional
  fun deleteProject(id: UUID) {
    projectRepository.lockOne(id).ifPresent { project ->
      val now = Instant.now(clock)
      project.deletedAt = now
      projectRepository.save(project)

      val things = thingRepository.findAll(id, Pageable.unpaged())
      things.content.forEach { thing ->
        thing.deletedAt = now
        thingRepository.save(thing)
      }
    }
  }

  private fun ProjectEntity.toDto() =
      ProjectDto(
          id = id,
          code = code,
          name = name,
          description = description,
          createdAt = createdAt,
          updatedAt = updatedAt,
          version = version,
      )
}

File: redfox-webapp/src/app/core/project/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Identity, PageResult } from '../common/common.model';
import { CreateProject, Project, UpdateProject } from './project.model';

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private readonly baseUrl = '/api/v1/projects';

  constructor(private readonly http: HttpClient) {}

  getProjects(page = 0, size = 20): Observable<PageResult<Project>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<PageResult<Project>>(this.baseUrl, { params });
  }

  getProject(id: string): Observable<Project> {
    return this.http.get<Project>(`${this.baseUrl}/${id}`);
  }

  createProject(request: CreateProject): Observable<Identity> {
    return this.http.post<Identity>(this.baseUrl, request);
  }

  updateProject(id: string, request: UpdateProject): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/${id}`, request);
  }

  deleteProject(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}
